File: src/main/java/pl/sdrozdz/fullstack/tester/TesterService.java
package pl.sdrozdz.fullstack.tester;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import pl.sdrozdz.fullstack.tester.dto.SearchCriteriaDto;
import pl.sdrozdz.fullstack.tester.dto.TesterDto;
import pl.sdrozdz.fullstack.tester.dto.TesterExperienceProjection;

import java.util.List;
import java.util.Optional;

import static java.util.stream.Collectors.toList;

@Service
@RequiredArgsConstructor
public class TesterService {

  private final TesterRepository testerRepository;

  public List<TesterDto> getTesters(Optional<SearchCriteriaDto> criteria) {
    List<Long> deviceIds = criteria.map(SearchCriteriaDto::getDevice).orElse(null);
    List<String> countries = criteria.map(SearchCriteriaDto::getCountry).orElse(null);
    return testerRepository
        .findByDevicesInAndCountryInOrderByExperienceDesc(deviceIds, countries)
        .stream()
        .map(this::convertToDto)
        .collect(toList());
  }

  private TesterDto convertToDto(TesterExperienceProjection projection) {
    final Tester tester = projection.getTester();
    return TesterDto.builder()
        .id(tester.getId())
        .firstName(tester.getFirstName())
        .lastName(tester.getLastName())
        .lastLogin(tester.getLastLogin())
        .country(tester.getCountry())
        .experience(projection.getExperience())
        .build();
  }
}

File: src/main/frontend/src/app/modules/tester/services/tester.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";

export interface SearchCriteriaDto {
  country: string[];
  device: number[];
}

export interface TesterDto {
  id: number;
  firstName: string;
  lastName: string;
  country: string;
  lastLogin: string;
  experience: number;
}

@Injectable()
export class TesterService {
  readonly API_URL = '/api/tester';

  constructor(private http: HttpClient) {
  }

  public getTesters(criteria: SearchCriteriaDto): Observable<TesterDto[]> {
    return this.http.post<TesterDto[]>(this.API_URL, criteria);
  }
}
